AK/MaybeOwned.h

```diff
@@ -32,6 +32,12 @@ class MaybeOwned {
     MaybeOwned(MaybeOwned&&) = default;
     MaybeOwned& operator=(MaybeOwned&&) = default;
 
+    template<DerivedFrom<T> U>
+    MaybeOwned(MaybeOwned<U>&& other)
+        : m_handle(downcast<U, T>(move(other.m_handle)))
+    {
+    }
```

Comment:
(nit: the 2nd term is missing a `.template get<NonnullOwnPtr<U>>()` after (other.m_handle))

The 2nd and 3rd subexpression of a a `?:` must have the same type, and that is deduced inside out. Since NNOPtr has an implicit conversion operator to pointer, this expression has the type `T*`. It creates a temporary NNOPtr, returns its ptr, and then destroys it.

It probably works if you throw in an explicit `Variant<NonnullOwnPtr<T>, T*>` ctor call in both the 2nd and 3rd subexpression, but at that point it's pretty wordy and imho worth of pulling into its own function?

I did replace the last line of `downcast` to say `return static_cast<NonnullOwnPtr<T>&&>(move(variant.template get<NonnullOwnPtr<U>>()));` instead of `return adopt_own<D>(*variant.template get<NonnullOwnPtr<U>>().leak_ptr());` (even though it's a bit longer, it's arguably cleaner and ownership is more solid).